File: InVehicleDevice/src/main/java/com/kogasoftware/odt/invehicledevice/view/fragment/utils/ScheduleUtil.java
package com.kogasoftware.odt.invehicledevice.view.fragment.utils;

import com.google.common.collect.Lists;
import com.kogasoftware.odt.invehicledevice.infra.contentprovider.table.OperationSchedule;
import com.kogasoftware.odt.invehicledevice.infra.contentprovider.table.PassengerRecord;

import java.io.Serializable;
import java.util.LinkedList;
import java.util.List;

public class ScheduleUtil implements Serializable {
    public static List<List<OperationSchedule>> getOperationSchedulesSortedPerPlatform(List<OperationSchedule> operationSchedules, List<PassengerRecord> passengerRecords) {
        List<List<OperationSchedule>> phaseOperationSchedulesList = Lists.newLinkedList();

        for (List<OperationSchedule> samePlatformOperationSchedules : getOperationScheduleListSamePlatformChunk(operationSchedules, passengerRecords)) {
            List<OperationSchedule> arrivalOperationSchedules = Lists.newArrayList();
            List<OperationSchedule> departureOperationSchedules = Lists.newArrayList();

            for (OperationSchedule operationSchedule : samePlatformOperationSchedules) {
                for (PassengerRecord passengerRecord : passengerRecords) {
                    if (passengerRecord.departureScheduleId.equals(operationSchedule.id) && !departureOperationSchedules.contains((operationSchedule))) {
                        departureOperationSchedules.add(operationSchedule);
                    } else if (passengerRecord.arrivalScheduleId.equals(operationSchedule.id) && !arrivalOperationSchedules.contains((operationSchedule))) {
                        arrivalOperationSchedules.add(operationSchedule);
                    }
                }
            }

            if (arrivalOperationSchedules.size() > 0) {
                phaseOperationSchedulesList.add(arrivalOperationSchedules);
            }
            if (departureOperationSchedules.size() > 0) {
                phaseOperationSchedulesList.add(departureOperationSchedules);
            }
        }
        return phaseOperationSchedulesList;
    }

    private static LinkedList<List<OperationSchedule>> getOperationScheduleListSamePlatformChunk(List<OperationSchedule> operationSchedules, List<PassengerRecord> passengerRecords) {

        boolean first = true;
        OperationSchedule previousOS = null;

        LinkedList<List<OperationSchedule>> platformOrderOperationScheduleLists = Lists.newLinkedList();

        for (OperationSchedule currentOS : operationSchedules) {
            if (first || !previousOS.platformId.equals(currentOS.platformId)) {
                List<OperationSchedule> samePlatformOperationSchedules = Lists.newArrayList();
                samePlatformOperationSchedules.add(currentOS);
                platformOrderOperationScheduleLists.add(samePlatformOperationSchedules);
                first = false;
            } else {
                platformOrderOperationScheduleLists.getLast().add(currentOS);
            }
            previousOS = currentOS;
        }
        return platformOrderOperationScheduleLists;
    }
}

File: InVehicleDevice/src/main/java/com/kogasoftware/odt/invehicledevice/view/fragment/utils/ScheduleUtil.kt
package com.kogasoftware.odt.invehicledevice.view.fragment.utils

import com.kogasoftware.odt.invehicledevice.infra.contentprovider.table.OperationSchedule
import com.kogasoftware.odt.invehicledevice.infra.contentprovider.table.PassengerRecord
import com.google.common.collect.Lists
import java.io.Serializable
import java.util.*

object ScheduleUtil : Serializable {
    @JvmStatic
    fun getOperationSchedulesSortedPerPlatform(operationSchedules: List<OperationSchedule>, passengerRecords: List<PassengerRecord>): List<List<OperationSchedule>> {
        val phaseOperationSchedulesList: MutableList<List<OperationSchedule>> = Lists.newLinkedList()
        for (samePlatformOperationSchedules in getOperationScheduleListSamePlatformChunk(operationSchedules, passengerRecords)) {
            val arrivalOperationSchedules: MutableList<OperationSchedule> = Lists.newArrayList()
            val departureOperationSchedules: MutableList<OperationSchedule> = Lists.newArrayList()
            for (operationSchedule in samePlatformOperationSchedules) {
                for (passengerRecord in passengerRecords) {
                    if (passengerRecord.departureScheduleId == operationSchedule.id && !departureOperationSchedules.contains(operationSchedule)) {
                        departureOperationSchedules.add(operationSchedule)
                    } else if (passengerRecord.arrivalScheduleId == operationSchedule.id && !arrivalOperationSchedules.contains(operationSchedule)) {
                        arrivalOperationSchedules.add(operationSchedule)
                    }
                }
            }
            if (arrivalOperationSchedules.size > 0) {
                phaseOperationSchedulesList.add(arrivalOperationSchedules)
            }
            if (departureOperationSchedules.size > 0) {
                phaseOperationSchedulesList.add(departureOperationSchedules)
            }
        }
        return phaseOperationSchedulesList
    }

    private fun getOperationScheduleListSamePlatformChunk(operationSchedules: List<OperationSchedule>, passengerRecords: List<PassengerRecord>): LinkedList<MutableList<OperationSchedule>> {
        var first = true
        var previousOS: OperationSchedule? = null
        val platformOrderOperationScheduleLists = Lists.newLinkedList<MutableList<OperationSchedule>>()
        for (currentOS in operationSchedules) {
            if (first || previousOS!!.platformId != currentOS.platformId) {
                val samePlatformOperationSchedules: MutableList<OperationSchedule> = Lists.newArrayList()
                samePlatformOperationSchedules.add(currentOS)
                platformOrderOperationScheduleLists.add(samePlatformOperationSchedules)
                first = false
            } else {
                platformOrderOperationScheduleLists.last.add(currentOS)
            }
            previousOS = currentOS
        }
        return platformOrderOperationScheduleLists
    }
}
